presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java

```diff
@@ -1102,27 +1105,28 @@ else if (criteria instanceof JoinOn) {
 
             for (Expression conjunct : ExpressionUtils.extractConjuncts((Expression) optimizedExpression)) {
                 conjunct = ExpressionUtils.normalize(conjunct);
-                if (!(conjunct instanceof ComparisonExpression)) {
-                    throw new SemanticException(NOT_SUPPORTED, node, "Non-equi joins not supported: %s", conjunct);
+                if (!isSupportedJoinConjunct(conjunct)) {
+                    throw new SemanticException(NOT_SUPPORTED, node, "Unsupported expression in join clause: %s", conjunct);
                 }
 
-                ComparisonExpression comparison = (ComparisonExpression) conjunct;
```

Comment:
Approach 1 seems like the right way to go to me, and I think I know why it's not working for you. It's slightly more complicated than I realized.
First, a bit of background:
The planner uses `TranslationMap` to convert field references into `Symbol`. You can see this in `TranslationMap.translateNamesToSymbols`. For this to work correctly, it needs the `analysis` to contain the correct field index into the relation. This whole process could be greatly simplified, but that's just how it works at the moment.

The reason that stopped working is that you analyzed the EQUALs expressions using the `output` relation type in `StatementAnalyzer`, but then the `TranslationMap` that gets used in the planner is from the left/right relation. To make this work I think you'll need to do the following, in both `StatementAnalyzer` and `RelationPlanner`:
* for EQUALs comparisons, use the `DependencyExtractor` to split the expression and then analyze/translate it using the left/right relation as appropriate
* for all other expressions (since they're going to be added as a post-JOIN filter step), analyze/translate them using the output relation from the JOIN. To do this in the planner it looks like you'll need to create a new `PlanBuilder` after this line `PlanNode root = new JoinNode(idAllocator.getNextId()` so that you can use that to rewrite the post-JOIN filter expressions. I think that will just work, but there may be some more issue there that we have to work through.

Let me know if you run into more issues!